### PR TITLE
feat(storage): report sst build memory for compute-node

### PR DESCRIPTION
**Pull request by Little-Wallace**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

We did not report  memory used by sstable-builder of compute-node, it made difficulty to understand which module use the most memory when compute-node crash because of OOM

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR contains user-facing changes.

<details><summary>Click here for Documentation</summary>

### Types of user-facing changes

Please keep the types that apply to your changes, and remove the others.

- Installation and deployment
- Connector (sources & sinks)
- SQL commands, functions, and operators
- RisingWave cluster configuration changes
- Other (please specify in the release note below)

### Release note

</details>

### CODE REVIEW

**Comment by Little-Wallace** [comment]
To avoid deadlock. I refactor the memory of output-sstable-builder and change it to `must require`. It means that although we can get the real value of memory usage, but it may cause OOM because we do not forbid memory allocate for sstable builder even if there is no avaliable memory left.


But it is still better than do nothing for sstable-builder. Because if the memory use more than quota, all operation of `ingest` will block actor and make backpressure to avoid add more data to hummock shared-buffer

File: src/storage/compactor/src/server.rs [hunk 1]
```diff
@@ -165,14 +165,11 @@ pub async fn compactor_serve(
     let observer_manager =
         ObserverManager::new_with_meta_client(meta_client.clone(), compactor_observer_node).await;
 
-    // use half of limit because any memory which would hold in meta-cache will be allocate by
-    // limited at first.
     let observer_join_handle = observer_manager.start().await;
-    let input_limit_bytes = compactor_memory_limit_bytes / 2;
+    // left some quota because we can not control it accurately.
+    let output_limit_bytes = compactor_memory_limit_bytes * 4 / 5;
```
Comment (Li0k) [inline src/storage/compactor/src/server.rs:170]: `compactor_memory_limit_bytes` is already reserved for memory during the calculation.

File: src/meta/src/hummock/compaction/picker/min_overlap_compaction_picker.rs [hunk 2]
```diff
@@ -23,6 +23,7 @@ use risingwave_pb::hummock::{InputLevel, Level, LevelType, SstableInfo};
 use super::{CompactionInput, CompactionPicker, LocalPickerStatistic};
 use crate::hummock::compaction::overlap_strategy::OverlapStrategy;
 use crate::hummock::level_handler::LevelHandler;
+pub const MAX_LEVEL_COUNT: usize = 32;
```
Comment (hzxa21) [inline src/meta/src/hummock/compaction/picker/min_overlap_compaction_picker.rs:26]: Will this be too small when ckpt_frequency=1?

File: src/storage/src/hummock/compactor/mod.rs [hunk 3]
```diff
@@ -249,6 +244,10 @@ impl Compactor {
                 return task_status;
             }
         }
+
+        let (input_memory, total_file_count, total_key_count) =
```
Comment (Li0k) [inline src/storage/src/hummock/compactor/mod.rs:248]: I think this limit is a bit too aggressive, for each merge_iter , each sst will only hold the memory of one block. Do we need to discuss how to evaluate the memory of a task? @hzxa21 @wenym1

Comment (hzxa21) [reply src/storage/src/hummock/compactor/mod.rs:248]: +1. I think we need to do experiment on the peak memory usage of a task with N SSTs in overlapping levels and M SSTs in non-overlapping levels.

File: src/storage/compactor/src/server.rs (hunk 1 above)
Comment (Little-Wallace) [reply src/storage/compactor/src/server.rs:170]: Oh. Just for safe... Maybe we can remove it in next PR

File: src/meta/src/hummock/compaction/picker/tier_compaction_picker.rs [hunk 4]
```diff
@@ -83,11 +84,16 @@ impl TierCompactionPicker {
 
             let mut compaction_bytes = level.total_file_size;
             let mut compact_file_count = level.table_infos.len() as u64;
+            // Limit sstable file count to avoid using too much memory.
```
[the remark this answers is not in the document]
Comment (Little-Wallace) [reply src/meta/src/hummock/compaction/picker/tier_compaction_picker.rs:87]: 32 sub-level is enough. It means that all data would be compact at most twice

File: src/config/example.toml [hunk 5]
```diff
@@ -83,7 +83,8 @@ cache_file_max_write_size_mb = 4
 [system]
 barrier_interval_ms = 1000
 checkpoint_frequency = 1
-sstable_size_mb = 256
+sstable_size_mb = 64
```
[the remark this answers is not in the document]
Comment (Little-Wallace) [reply src/config/example.toml:86]: To reduce the memory use by builder  in compute-node.

**Comment by Little-Wallace** [comment]
> +1. I think we need to do experiment on the peak memory usage of a task with N SSTs in overlapping levels and M SSTs in non-overlapping levels.

Agree, We need more experiment to explain why compact task use so much memory when the number of levels is too large. This PR is just to reduce the probablity of OOM. It does not solve the problem, and we will investigate on it continue.

**Pull request closed without merging**